ext/semian/tickets.c

```diff
@@ -27,31 +30,101 @@ update_ticket_count(update_ticket_count_t *tc)
 
   return Qnil;
 }
+
+int
+update_tickets_from_quota(int sem_id, double quota)
+{
+  int delta = 0;
+  int tickets = 0;
+  int state = 0;
+  update_ticket_count_t tc;
+
+  // DEBUG PRINT - delete me
+  //printf("%d %d %d %d\n",
+  //  get_sem_val(sem_id, SI_SEM_TICKETS),
+  //  get_sem_val(sem_id, SI_SEM_CONFIGURED_TICKETS),
+  //  get_sem_val(sem_id, SI_SEM_REGISTERED_WORKERS),
+  //  get_sem_val(sem_id, SI_SEM_CONFIGURED_WORKERS)
+  //);
+  // If the configured worker count doesn't match the registered worker count, adjust it.
+  // and adjust the underlying tickets available to match.
+  sem_meta_lock(sem_id);
```

Comment:
Lock placed here as this is the only critical section.

It's important that the lock is *before* the delta is computed, so that multiple processes don't waste their time here if they don't need to on the perform_semop.